src/model.ts

```diff
@@ -21,6 +21,7 @@ class Model<T> extends EventEmitter {
   Document: { new<T>(data: T): Document<T> };
   Query: { new<T>(data: Document<T>[]): Query<T> };
   _database: Database;
+  [key : string]: any;
```

Comment:
https://github.com/hexojs/warehouse/blob/af981d3bf003c3bf0f5dad9f3861ed39ee0648fc/src/model.ts#L79-L82
Solve the above code, but as with #252, it's not a good solution. Model knows nothing about static and method types in Schema.
@SukkaW Do you have any good solutions?

Reply:
I was going to add another generic for setting the type of static and method, but its not possible to act directly on the class.
```ts
interface Extra {
  statics: Record<string, (...args: any[]) => any>;
  methods: Record<string, (...args: any[]) => any>;
}

type AddThis<R extends Record<string, (...args: any[]) => any>, T> = {
  [K in keyof R]: (this: T, ...args: Parameters<R[K]>) => ReturnType<R[K]>;
}

class Schema<T = any, K extends Partial<Extra> = Extra> {
  statics!: K['statics'] extends Record<string, (...args: any[]) => any> ? AddThis<K['statics'], T> : Record<string, (this: T, ...args: any[]) => any>;
  methods!: K['methods'] extends Record<string, (...args: any[]) => any> ? AddThis<K['methods'], T> : Record<string, (this: T, ...args: any[]) => any>;

}

interface AssetSchema {
  _id: string
}

const a = new Schema<AssetSchema, {
  statics: {
    add: (a: number, b: number) => number;
  }
}>().statics.add // <-- (this: AssetSchema, a: number, b: number) => ReturnType<(a: number, b: number) => number>

const b = new Schema<AssetSchema>().methods // <-- AddThis<Record<string, (...args: any[]) => any>, AssetSchema>

class Model<T, K extends Partial<Extra> = Extra> {
  schema!: Schema<T, K>
  // [key in keyof K['statics']]: K['statics'][key] illegal
}
```